torch/csrc/api/include/torch/detail/any.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <torch/csrc/utils/memory.h>
+#include <torch/csrc/utils/variadic.h>
+
+#include <ATen/Error.h>
+
+#include <memory>
+#include <typeinfo>
+#include <utility>
+
+namespace torch {
+class Any;
+namespace detail {
+template <typename T>
+Any make_any(T&& value);
+} // namespace detail
+} // namespace torch
+
+namespace torch {
+/// A simplified implementation of `std::any` or `boost::any` which stores a
+/// type erased object, whose concrete value can be retrieved at runtime by
+/// checking if the `typeid()` of a requested type matches the `typeid()` of the
+/// object stored. It is simplified in that it does not handle copying, as we do
+/// not require it for our use cases. Moves are sufficient.
+class Any {
+ public:
+  /// Default construction is disallowed. Thus our invariant: an `Any` contains
+  /// an object at all times, from its construction, to its destruction.
+  Any() = delete;
+
+  /// Move construction and assignment is allowed, and follows the default
+  /// behavior of move for `std::unique_ptr`.
+  Any(Any&&) = default;
+  Any& operator=(Any&&) = default;
+
+  /// Copy is disallowed, because we don't need it.
+  Any(const Any& other) = delete;
+  Any& operator=(const Any& other) = delete;
+
+  /// Returns the value contained in the `Any` if the type passed as template
+  /// parameter matches the type of the object stored, and returns a null
+  /// pointer otherwise.
+  template <typename T>
+  T* try_get() {
```

Comment:
Confused about ownership. Does this return a non-owning pointer? Then the dereference in `get()` seems to cause a copy to happen. Is this intentional?